File: Pollock/TypeConverters.cs
using Pollock.Infrastructure;
using System;
using System.Collections.Generic;

namespace Pollock
{
    public class TypeConverters
    {
        private static IDictionary<Type, ITypeConverter> converters =
            new Dictionary<Type, ITypeConverter>();

        private static void RegisterEnumWithMemberRender<T>()
        {
            Register<T>(new EnumWithMemberRenderConverter<T>());
        }

        private static void Register<T>(ITypeConverter converter) 
        {
            converters[typeof(T)]= converter;
        }

        public static ITypeConverter Get<T>() where T: struct
        {
            var type = typeof(T);
            if (!converters.ContainsKey(type))
            {
                RegisterEnumWithMemberRender<T>();
            }
            return converters[type];
        }
    }
}

File: PollockFs/TypeConverters.fs
namespace Pollock

open System
open System.Collections.Generic

module public TypeConverters = 
    let cs = new Dictionary<Type, ITypeConverter>()
    
    let register<'T> (converter : ITypeConverter) = 
        let t = typeof<'T>
        if (not(cs.ContainsKey(t))) then
            cs.Add(t, converter)
        else
            cs.[t] <- converter 
    
    let registerEnumWithMemberRender<'T when 'T : struct>() = register<'T> (new EnumWithMemberRenderConverter<'T>())
    
    let get<'T when 'T : struct>() : ITypeConverter = 
        let t = typeof<'T>
        if (not (cs.ContainsKey(t))) then registerEnumWithMemberRender<'T>() |> ignore
        else ()
        cs.[t]
